### ISSUE TITLE
talonfmt emits CRLF newlines on Windows instead of preserving the original ones

### ISSUE THREAD

**Issue by saidelike** [wenkokke/talonfmt#119]
I have a hook for talonfmt only

```
> more .pre-commit-config.yaml
minimum_pre_commit_version: "3.2.0"
repos:
  - repo: https://github.com/wenkokke/talonfmt
    rev: 1.10.1
    hooks:
      - id: talonfmt
        args: ["--in-place"]
```

# test 1

i have global core.autocrlf = false and I am on Windows.

```
>git config --global core.autocrlf
false

>git config --local core.autocrlf
```

pre-commit fails:

```
>pre-commit run --all-files
talonfmt.................................................................Failed
- hook id: talonfmt
- files were modified by this hook
```

All my talon files were changed from LF to CRLF.

```
>git status
On branch main
Your branch is up to date with 'origin/main'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
        modified:   modes/mode_command.talon
        modified:   modes/mode_normal.talon
        modified:   modes/mode_replace.talon
        modified:   modes/mode_terminal.talon
        modified:   modes/mode_visual.talon
        ...
>git diff
diff --git a/modes/mode_normal.talon b/modes/mode_normal.talon
index c3908d5..a65e37b 100644
--- a/modes/mode_normal.talon
+++ b/modes/mode_normal.talon
@@ -1,11 +1,11 @@
-win.title: /VIM MODE:n/
-win.title: /VIM MODE:no/
-win.title: /VIM MODE:nov/
-win.title: /VIM MODE:noV/
-win.title: /VIM MODE:no^V/
-win.title: /VIM MODE:niI/
-win.title: /VIM MODE:niR/
-win.title: /VIM MODE:niV/
-win.title: /VIM MODE:nt/
--
-tag(): user.vim_normal_mode
+win.title: /VIM MODE:n/^M
+win.title: /VIM MODE:no/^M
+win.title: /VIM MODE:nov/^M
+win.title: /VIM MODE:noV/^M
+win.title: /VIM MODE:no^V/^M
+win.title: /VIM MODE:niI/^M
+win.title: /VIM MODE:niR/^M
+win.title: /VIM MODE:niV/^M
+win.title: /VIM MODE:nt/^M
+-^M
+tag(): user.vim_normal_mode^M
```

# test 2

If I revert everything and if I set core.autocrlf to true, the problem does not appear:

```
>git config --global core.autocrlf
true

>git config --local core.autocrlf
```

```
>pre-commit run --all-files
talonfmt.................................................................Passed
```

**Comment by saidelike** [comment]
I am wondering if the problem is due to
https://github.com/wenkokke/talonfmt/blob/15e21fcc107774fa703d83d04508cdff5f7ef4b7/src/talonfmt/cli.py#L172

**Comment by saidelike** [comment]
You read the file as binary

https://github.com/wenkokke/talonfmt/blob/15e21fcc107774fa703d83d04508cdff5f7ef4b7/src/talonfmt/cli.py#L129

 but write it as non-binary

https://github.com/wenkokke/talonfmt/blob/15e21fcc107774fa703d83d04508cdff5f7ef4b7/src/talonfmt/cli.py#L172

**Comment by wenkokke** [comment]
Not sure if that's the problem. Tree sitter requires you read the file as binary, but after parsing the contents are text.

**Comment by saidelike** [comment]
Interesting. How do you preserve new lines types then? I guess if there are mixed CR and CRLF, you would need to not change them as it is not the role of talonfmt? Same thing if there are only CR. Same thing if there are only CRLF.

I forgot to mention that in test2 case, even if no error, you still changed from CR to CRLF.

**Comment by wenkokke** [comment]
There's a good chance I don't even attempt to preserve newlines. I think I just emit the standard newline for the current platform.

EDIT: Nope, I just normalize all newlines to LF, so I'm not entirely sure where the CR even comes from.

**Comment by saidelike** [comment]
Well the standard newlines for Windows are CRLF. And yes i can confirm that you always emit CRLF on Windows. This is the problem. If the original file had LF newlines, which is possible on Windows as most editors support it too, you basically changed the original newlines. 

My understanding is opening the file with "w" on Windows and writing lines to it will create Windows newlines so CRLF. 

But i don't understand enough about `talonfmt()` to know if you emit the platform specific newlines there.

**Comment by saidelike** [comment]
Indeed, even if talonfmt() creates LF, opening the file with "w" (instead of "wb") will convert them in CRLF :D

```
>>> open("test", "w").write("hello\nworld\ntalonfmt")
20
>>> open("test", "rb").read()
b'hello\r\nworld\r\ntalonfmt'
```

**Comment by wenkokke** [comment]
> Indeed, even if talonfmt() creates LF, opening the file with "w" (instead of "wb") will convert them in CRLF :D
> 
> 
> 
> ```
> 
> >>> open("test", "w").write("hello\nworld\ntalonfmt")
> 
> 20
> 
> >>> open("test", "rb").read()
> 
> b'hello\r\nworld\r\ntalonfmt'
> 
> ```

Wait, what? How? Why?

**Comment by saidelike** [comment]
> In text mode, the default when reading is to convert platform-specific line endings (\n on Unix, \r\n on Windows) to just \n. When writing in text mode, the default is to convert occurrences of \n back to platform-specific line endings. This behind-the-scenes modification to file data is fine for text files, but will corrupt binary data like that in JPEG or EXE files. Be very careful to use binary mode when reading and writing such files.

From https://docs.python.org/3/tutorial/inputoutput.html#reading-and-writing-files
